### DISCUSSION TITLE
Guide about InputControl

### DISCUSSION THREAD

**Discussion by linonetwo** [retejs/rete#619] [Ideas]
Is InputControl the only type of control?

I can't find a page about Control in the new doc, apart from https://retejs.org/examples/react/controls

This is the code snippet from import/export example, so `reteType` field is unnecessary, 

```ts
export function serializeControl(control: ClassicPreset.Control) {
  if (control instanceof ClassicPreset.InputControl) {
    return {
      reteType: 'ClassicPreset.InputControl' as const,
      id: control.id,
      readonly: control.readonly,
      type: control.type as string,
      value: control.value as unknown,
    };
  }
}
```

And when deserializing from JSON, we use `control.type` to decide which control we use in the UI.

For example:

```
const availableControls = {
  inputControl: new ClassicPreset.InputControl("number", {
    initial: 0,
    change(value) {
      progressControl.percent = value;
      area.update("control", progressControl.id);
    }
  });
}

a.addControl(control.type, availableControls[control.type]);
```

**Comment by Ni55aN** [comment]
> Is InputControl the only type of control?

yes, it serves as a model to separate data/state from presentation

https://retejs.org/docs/guides/renderers/react#controls

while the `ClassicPreset` is a part of `rete` package

**Reply by linonetwo** [reply]
Is there an example about how to update control? For example, when in multiplayer mode we receive latest control value from other client:

```ts
export function applyNodeDataToReteNode(node: ClassicPreset.Node, config: INodeExportedJSON) {
  for (const controlName in node.controls) {
    const control = node.controls[controlName];
    if (control !== undefined) {
      control // ?
    }
  }
```

**Reply by linonetwo** [reply]
I know, we should cast it to InputControl:

```ts
export function applyNodeDataToReteNode(node: ClassicPreset.Node, config: INodeExportedJSON) {
  for (const controlName in node.controls) {
    const control = node.controls[controlName] as ClassicPreset.InputControl<'number' | 'text'>;
    control.setValue(config.config[controlName] as string | number);
  }
}
```

**Reply by Ni55aN** [reply]
https://retejs.org/docs/faq/#force-update
